src/sage/modules/free_module_pseudohomspace.py

```diff
@@ -0,0 +1,158 @@
+
+# ****************************************************************************
+#  Copyright (C) 2024 Yossef Musleh <jbobicus@gmail.com>
+#
+#  Distributed under the terms of the GNU General Public License (GPL)
+#
+#    This code is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty
+#    of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+#  See the GNU General Public License for more details; the full text
+#  is available at:
+#
+#                  https://www.gnu.org/licenses/
+# ****************************************************************************
+import sage.categories.homset
+from sage.structure.element import is_Matrix
+from sage.matrix.constructor import matrix, identity_matrix
+from sage.matrix.matrix_space import MatrixSpace
+from sage.misc.cachefunc import cached_method
+from sage.categories.morphism import Morphism
+from sage.misc.lazy_import import lazy_import
+
+lazy_import('sage.rings.derivation', 'RingDerivation')
+
+class FreeModulePseudoHomspace(sage.categories.homset.HomsetWithBase):
+    r"""
+    This class implements the space of Pseudomorphisms with a fixed twist.
+
+    For free modules, the elements of a pseudomorphism correspond to matrices
+    which define the mapping on elements of a basis.
+
+    EXAMPLES::
+
+        sage: F = GF(125); M = F^2; twist = F.frobenius_endomorphism()
+        sage: PHS = M.PseudoHom(twist)
+        sage: h = PHS([[1, 2], [1, 1]])
+        sage: e = M((4*F.gen()^2 + F.gen() + 2, 4*F.gen()^2 + 4*F.gen() + 4))
+        sage: h(e)
+        (3*z3^2 + z3, 4*z3^2 + 3*z3 + 3)
+
+    """
+    def __init__(self, X, Y, twist=None):
+        self._domain = X
+        self._codomain = X
+        if Y is not None:
+            self._codomain = Y
+        self.base_homspace = self._domain.Hom(self._codomain)
+        if twist is None:
+            return
+        if (twist.domain() is not self.domain().coordinate_ring()
+          or twist.codomain() is not self.codomain().coordinate_ring()):
+            raise TypeError("twisting morphism domain/codomain do not match\
+                            coordinate rings of the modules")
+        elif isinstance(twist, Morphism) or isinstance(twist, RingDerivation):
+            self.twist = twist
+        else:
+            raise TypeError("twist is not a ring morphism or derivation")
+
+    def __call__(self, A, **kwds):
```

Comment:
I think this code should go in the `__init__` function of the element.
And you need to use the `_element_constructor` stuff.